### PR TITLE
[JIT] Remove profile nodes before BatchMM.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44073 [TensorExpr] Fuser: do not fuse ops with 0-dim tensors.
* **#43961 [JIT] Remove profile nodes before BatchMM.**

Currently we're removing prim::profile nodes and embed the type info
directly in the IR right before the fuser, because it is difficult to
fuse in a presence of prim::profile nodes. It turns out that BatchMM has
a similar problem: it doesn't work when there are prim::profile nodes in
the graph. These two passes run next to each other, so we could simply
remove prim::profile nodes slightly earlier: before the BatchMM pass.

Differential Revision: [D23453266](https://our.internmc.facebook.com/intern/diff/D23453266)

### CODE REVIEW

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 1]
```diff
@@ -224,7 +224,12 @@ void runDiffGraphPasses(std::shared_ptr<Graph>& graph) {
     // TupleConstruct / TupleUnpack pairs can still be present at this point
     // and must be removed for fusion.
     LowerSimpleTuples(graph);
-    GRAPH_DUMP("After LowerSimpleTuples, before BatchMM", graph);
+    GRAPH_DUMP(
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:227]: Can you add a comment abt why we're calling `RemoveProfileNodesAndSpecializeTypes`, and that any pass that uses specialized tensor types has to be run after  `FuseTensorExprs`.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 2]
```diff
@@ -224,7 +224,12 @@ void runDiffGraphPasses(std::shared_ptr<Graph>& graph) {
...
+        "After LowerSimpleTuples, before RemoveProfileNodesAndSpecializeTypes",
+        graph);
+    RemoveProfileNodesAndSpecializeTypes(graph);
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:230]: Can we call `EraseShapeInformation` if `!tensorExprFuserEnabled()`

**Review by eellison — APPROVED** [review]
LGTM

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:249]: I moved here a call to `RemoveTensorTypeSpecializations` from the TE fuser - that should be equivalent, right?

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Pull request closed without merging**